### ISSUE TITLE
Behavior change in `std.testing.refAllDecls()`

### ISSUE THREAD

**Issue by marnix** [ziglang/zig#10848]
### Zig Version

0.10.0-dev.560+01d48e55a

### Steps to Reproduce

Run both Zig 0.9.x and a recent (Feb 9, 2022 or later) master Zig

```
$ zig version
0.9.0
$ zig test a.zig
All 2 tests passed.
$ zig-recent version
0.10.0-dev.560+01d48e55a
$ zig-recent test a.zig
All 1 tests passed.
$ 
```
with the files below, and compare the results: The test in c.zig is suddenly silently skipped.

a.zig:
```zig
const std = @import("std");

test {
    std.testing.refAllDecls(@import("b.zig"));
}
```

b.zig:
```zig
const c = @import("c.zig");

pub fn functionB() void {
    c.functionC();
}
```

c.zig:
```zig
pub fn functionC() void {}

test {}
```

### Expected Behavior

Zig master should also run the test in c.zig.

(Or the Zig master documentation should be updated to make clear which declarations are and which are not referenced.)

### Actual Behavior

Zig master skips the test in c.zig, which is only referenced indirectly by a.zig.

**Comment by marnix** [comment]
I may be mistaken, but the cause may be https://github.com/ziglang/zig/commit/75bbc74d56f26da812c47436c6f5712d3cfde12d#diff-c71e716cab4f75e8fa5debd27c419b3183c1b7bdd4768257d2cf4a37a57a6703, or something else in that PR.  Comparing two GitHub actions of mine ([one](https://github.com/marnix/zigmmverify/runs/5000953668?check_suite_focus=true#step:5:4) and [two](https://github.com/marnix/zigmmverify/runs/5123447090?check_suite_focus=true#step:5:4)), it seems the change occurred between January 31 and February 9

Up until January 31, those tests-from-imports were referenced and run, but since today February 9 they aren't anymore.

My main worry, I guess, is that if my code is suddenly silently skipping tests, other's code can do the same, including stage2 stuff...

**Comment by matu3ba** [comment]
Did you run `git bisect` with continuously building the compiler to figure out when the problem was introduced?

**Comment by matu3ba** [comment]
The first commit, where your program compiles and runs with but has weirdly looking output (I would expect disabled logging unless there is a failing test) is 8a97807d6812f62db4c3088fecd04a98a84b9943.
The output is 
```txt
[user@pc zigmmverify]$ zig build
Test [1/75] verify.test "proof with $d violation"... [default] (warn): $d Q R $. expected but not found in the following list:

[default] (warn): (end of list)

[default] (warn): 
Error error.DVRMissing happened...

Test [73/75] prove.test "compare unequal expressions"... 
eqExpr: actual = a $x, expected = b $x.
Test [74/75] prove.test "compare unequal expressions, constant vs variable"... 
eqExpr: actual = class $x, expected = class x.
All 75 tests passed.
```

**Comment by matu3ba** [comment]
I started the binary search from 8a97807d6812f62db4c3088fecd04a98a84b9943 (Full response file (*.rsp) support) towards f0400ad93eac984332c938b39e9c1627062d6b6a.

d3e40b0d8085ec82e6e5b33f2010de6cb9779ce5 (Merge pull request #10743 from m-radomski/master)
is the last good commit.

aa326328d0b4a0d4ee8167be9b2ab25fd560b0c9 (stage1: remove the `data` field from TypeInfo.Declaration)
broke the build in the test command with
```txt
[user@pc zigmmverify]$ rm -fr zig-cache/ zig-out/ "${HOME}"/.cache/zig/ && zig build
/home/misterspoon/dev/git/zig/zig/master/lib/std/testing.zig:468:12: error: unable to evaluate constant expression
    inline for (std.meta.declarations(T)) |decl| {
           ^
./src/alltests.zig:24:28: note: called from here
    std.testing.refAllDecls(@import("main.zig"));
                           ^
./src/alltests.zig:23:9: note: called from here
test "" {
```

**Comment by matu3ba** [comment]
From looking at the remaining commit this was a branch merge in 0298442100b2d5707099f09d29af554e8c6ac87e
of 4d22fa5a2a4afce59b91efc7a6142ec633f0bd4d, 75bbc74d56f26da812c47436c6f5712d3cfde12d and aa326328d0b4a0d4ee8167be9b2ab25fd560b0c9

**Comment by nektro** [comment]
`c` in `b.zig` isnt `pub`

**Comment by Vexu** [comment]
My commit in https://github.com/ziglang/zig/pull/10849 should fix this.

**Comment by ominitay** [comment]
> `c` in `b.zig` isnt `pub`

Making `c` `pub` doesn't fix this.

**Comment by Vexu** [comment]
Fixed by 8937f18a6f8496e011b13cb086b7948b5f1d540e

**Comment by marnix** [comment]
Thanks @Vexu! I can confirm that my project started to run all tests again Tue, 15 Feb 2022 03:34:44 GMT (it last ran just 1 test Mon, 14 Feb 2022 03:35:12 GMT), which matches the 3eb29f15 = #10849 commit time of 2022-02-14T10:34:33Z.
